### ISSUE TITLE
How to decode namerecord?

### ISSUE THREAD

**Issue by moi15moi** [MicrosoftDocs/typography-issues#956]
### Quote from the documentation
> All string data for platform 3 must be encoded in UTF-16BE.
Note: Some legacy CJK fonts had non-conforming string data using mixed encoding schemes. For example, the MingLi font included in the Traditional Chinese edition of Windows 95 had name entries for platform 3, encoding ID 4 (Big5), and encoded Latin strings using UTF-16BE but Traditional Chinese strings using code page 950.

### Font MingLi that is been mentionned in the documentation
I downloaded the font from this [.iso](https://archive.org/details/Win95CCHT)

If you need the font, you can download it [here](https://mega.nz/file/zh0CWC6Y#9gWeFf4vqaPubBsCNgoYYyYBDcypPJckGEMY0iGJ6FU)
I attached the font only to make the life easier of the people who gonna answer me. If it is illegal, please warn me.

### Question
When I try to decode all the record with an ``platEncID = 4`` with ``utf_16_be``, no exception are raised.
So, how are we supposed to know we need to use ``cp950`` for X record?

Also, from what I can see, decoding it with ``big5`` return the same result has with ``cp950``. It this just a pure coincidence?

### Name record decoded with [fonttools](https://github.com/fonttools/fonttools)
- Decoded with utf_16_be:
```xml
    <namerecord nameID="0" platformID="3" platEncID="4" langID="0x404">
      (c) Copyright CompuFont Ltd. 1992
    </namerecord>
    <namerecord nameID="1" platformID="3" platEncID="4" langID="0x404">
      닓ꧺ엩
    </namerecord>
    <namerecord nameID="2" platformID="3" platEncID="4" langID="0x404">
      Regular
    </namerecord>
    <namerecord nameID="3" platformID="3" platEncID="4" langID="0x404">
      닓ꧺ엩
    </namerecord>
    <namerecord nameID="4" platformID="3" platEncID="4" langID="0x404">
      닓ꧺ엩
    </namerecord>
    <namerecord nameID="5" platformID="3" platEncID="4" langID="0x404">
      1 May, 1992; 1.00, initial release
    </namerecord>
    <namerecord nameID="6" platformID="3" platEncID="4" langID="0x404">
      MingLi43
    </namerecord>
    <namerecord nameID="7" platformID="3" platEncID="4" langID="0x404">
      Trademark of CompuFont Ltd., HK
    </namerecord>
```

- Decoded with code page 950:
```xml
    <namerecord nameID="0" platformID="3" platEncID="4" langID="0x404">
      (c) Copyright CompuFont Ltd. 1992
    </namerecord>
    <namerecord nameID="1" platformID="3" platEncID="4" langID="0x404">
      細明體
    </namerecord>
    <namerecord nameID="2" platformID="3" platEncID="4" langID="0x404">
      Regular
    </namerecord>
    <namerecord nameID="3" platformID="3" platEncID="4" langID="0x404">
      細明體
    </namerecord>
    <namerecord nameID="4" platformID="3" platEncID="4" langID="0x404">
      細明體
    </namerecord>
    <namerecord nameID="5" platformID="3" platEncID="4" langID="0x404">
      1 May, 1992; 1.00, initial release
    </namerecord>
    <namerecord nameID="6" platformID="3" platEncID="4" langID="0x404">
      MingLi43
    </namerecord>
    <namerecord nameID="7" platformID="3" platEncID="4" langID="0x404">
      Trademark of CompuFont Ltd., HK
    </namerecord>
```

- Decoded with big5
```xml
    <namerecord nameID="0" platformID="3" platEncID="4" langID="0x404">
      (c) Copyright CompuFont Ltd. 1992
    </namerecord>
    <namerecord nameID="1" platformID="3" platEncID="4" langID="0x404">
      細明體
    </namerecord>
    <namerecord nameID="2" platformID="3" platEncID="4" langID="0x404">
      Regular
    </namerecord>
    <namerecord nameID="3" platformID="3" platEncID="4" langID="0x404">
      細明體
    </namerecord>
    <namerecord nameID="4" platformID="3" platEncID="4" langID="0x404">
      細明體
    </namerecord>
    <namerecord nameID="5" platformID="3" platEncID="4" langID="0x404">
      1 May, 1992; 1.00, initial release
    </namerecord>
    <namerecord nameID="6" platformID="3" platEncID="4" langID="0x404">
      MingLi43
    </namerecord>
    <namerecord nameID="7" platformID="3" platEncID="4" langID="0x404">
      Trademark of CompuFont Ltd., HK
    </namerecord>
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9bc5a9d9-70e1-97b3-6926-c176e1a2401d
* Version Independent ID: 1afb0f5b-d1a7-7d69-d175-a5c9382b74be
* Content: [name - Naming table (OpenType 1.9) - Typography](https://docs.microsoft.com/en-us/typography/opentype/spec/name#platform-specific-encoding-and-language-ids-windows-platform-platform-id-3)
* Content Source: [typographydocs/opentype/spec/name.md](https://github.com/MicrosoftDocs/typography/blob/live/typographydocs/opentype/spec/name.md)
* Product: **typography**
* GitHub Login: @PeterCon
* Microsoft Alias: **alib**

**Comment by PeterCon** [comment]
>So, how are we supposed to know we need to use cp950 for X record?

I'm not sure there is a deterministic way that would be 100% reliable. There may be some strings that have a valid interpretation in Big5/cp950 but are not valid in UTF-16BE. On the other hand, there may be strings that have a valid interpretation both in Big5/cp950 and also in UTF-16BE, but that are gibberish in UTF-16BE. (You appear to have demonstrated just that, above.)

>Also, from what I can see, decoding it with big5 return the same result has with cp950. It this just a pure coincidence?

No, as far as I know, cp950 is the same as Big5.

**Comment by moi15moi** [comment]
>I'm not sure there is a deterministic way that would be 100% reliable.

Then, it is possible to share how windows does it? 
If all program use the same method, it would allow a far more consistent results across software.

For reference:
- Big5: http://www.unicode.org/Public/MAPPINGS/OBSOLETE/EASTASIA/OTHER/BIG5.TXT
- CP950: http://www.unicode.org/Public/MAPPINGS/VENDORS/MICSFT/WINDOWS/CP950.TXT

**Comment by PeterCon** [comment]
Windows GDI has special case handling for subfamily name (name ID 2) that checks for Wansung (3/5) and Big5 (3/4) with a comment, "MingLi.TTF's bug, Style use Unicode encoding, not BIG5 encoding".

I don't think it's a good idea to spec out a behaviour for fonts considered non-conformant. That would be, in effect, establishing interoperability for data that previously was considered non-conformant, and in the process effectively paving the way for it to be conformant. And especially when one platform added special handling for one particular font that it happened to ship.

There are other messy areas in which fonts have done things that are contrary to expectation: head.fontRevision, OS/2 usWeightClass and usWidthClass... I don't think the OT spec should try to document the different ways these fields have been set in practice, let alone spec out how implementations should parse them.

**Comment by moi15moi** [comment]
Ok, so in brief, microsoft recommand to decode all name from platform 3 with `utf16be`, but with some really old cjk font (and not even all of them), the result can be wrong.

**Comment by PeterCon** [comment]
Since you were asking, I did a little investigation and discovered that Windows doesn't do things quite as I described. Moreover, there's a slight difference between GDI and DWrite: apparently DWrite isn't making allowance for the quirk that existed in the Win95 fonts (not really an issue since DWrite doesn't run on Win95).

For 3/3 ("PRC" / GB2312 / cp936), 3/4 (Big5 / cp950) and 3/5 ("Wansung" / cp949), both GDI and DWrite assume the encoding of name strings is in those encodings—with GDI (but not DWrite) having an exception due to the Win95 font quirk: for the subfamily string (name ID 2), GDI always assumes UTF-16BE. For all other encoding IDs, UTF-16BE is assumed in both GDI and DWrite.

Summarizing, here's pseudo code describing GDI logic:

```
If Windows platform:
    If GB: string is GB
    Elseif Big5 || Wansung:
        If subfamily: string is UTF-16BE
        Else: string is per encoding ID
    Else: string is UTF-16BE
```

And here's pseudo code describing DWrite logic:

```
If Windows platform:
    If GB || Big5 || Wansung: string is per encoding ID
    Else: string is UTF-16BE
```

So, the current statement in the 'name' chapter (always UTF-16BE) isn't quite correct. I'll need to think of how to improve since, for any new fonts, none of the legacy encodings should be used in either 'cmap' or 'name'.

**Comment by moi15moi** [comment]
Wow, thanks for this awesome answer!

Just to be sure I correctly understood, when you assumed the encoding like this: ``If GB: string is GB``, you mean the encoding_id is equal to 3 which is PRC?

It just confused me since there is no encoding_id for GB: https://docs.microsoft.com/en-us/typography/opentype/spec/name#windows-encoding-ids

**Comment by rcombs** [comment]
Meanwhile, it looks like SJIS (CP932, encoding_id 2) is probably supported in mapping tables, but not in names?

**Comment by PeterConstable** [comment]
Sorry, "GB" = "PRC". (China character set standards have designations that begin "GB".)

>Meanwhile, it looks like SJIS (CP932, encoding_id 2) is probably supported in mapping tables, but not in names?

Correct.

All of the platform 3 encoding IDs are supported in 'cmap' as described. E.g., a 3/4 'cmap' subtable is, in fact, Big5 (cp 950) encoding, and would likely be using a format 2 subtable. Apps should support those encodings for 'cmap' subtables (though new fonts should only use Unicode 'cmap' subtables.)

But in the 'name' table, strings with 3/0 (symbol), 3/2 (ShiftJIS) or 3/6 (Johab) name records should be encoded in UTF-16, in spite of the encoding ID.

**Comment by PeterCon** [comment]
See the [OT 1.9.1 alpha](https://docs.microsoft.com/en-us/typography/opentype/otspec191alpha/ot191alpha) for draft revisions addressing this issue.

**Comment by PeterCon** [comment]
This is fixed in OT 1.9.1. See the [OT 1.9.1 beta](https://docs.microsoft.com/en-us/typography/opentype/otspec191/ot191beta) for draft revisions addressing this issue.
